File: lib/algorithms.ts
import { Location, Circle } from '@/types';

/**
 * Convert lat/lng to unit vector on sphere
 */
function latLngToVector(lat: number, lng: number): [number, number, number] {
  const latRad = (lat * Math.PI) / 180;
  const lngRad = (lng * Math.PI) / 180;
  return [
    Math.cos(latRad) * Math.cos(lngRad),
    Math.cos(latRad) * Math.sin(lngRad),
    Math.sin(latRad),
  ];
}

/**
 * Convert unit vector back to lat/lng
 */
function vectorToLatLng(v: [number, number, number]): { lat: number; lng: number } {
  const length = Math.sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2]);
  const x = v[0] / length;
  const y = v[1] / length;
  const z = v[2] / length;

  const lat = (Math.atan2(z, Math.sqrt(x * x + y * y)) * 180) / Math.PI;
  const lng = (Math.atan2(y, x) * 180) / Math.PI;

  return { lat, lng };
}

/**
 * Compute centroid on sphere
 * Avoids artifacts near poles & 180° crossings
 */
export function computeCentroid(locations: Location[]): { lat: number; lng: number } | null {
  if (locations.length === 0) return null;

  // Convert all locations to unit vectors and average
  let sumX = 0,
    sumY = 0,
    sumZ = 0;

  for (const loc of locations) {
    const [x, y, z] = latLngToVector(loc.lat, loc.lng);
    sumX += x;
    sumY += y;
    sumZ += z;
  }

  const avgVector: [number, number, number] = [
    sumX / locations.length,
    sumY / locations.length,
    sumZ / locations.length,
  ];

  return vectorToLatLng(avgVector);
}

/**
 * Calculate distance between two points on Earth using Haversine formula
 * Returns distance in meters
 */
export function haversineDistance(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const R = 6371000; // Earth's radius in meters
  const φ1 = (lat1 * Math.PI) / 180;
  const φ2 = (lat2 * Math.PI) / 180;
  const Δφ = ((lat2 - lat1) * Math.PI) / 180;
  const Δλ = ((lng2 - lng1) * Math.PI) / 180;

  const a =
    Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
    Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c;
}

/**
 * Compute circle from 2 points (diameter)
 */
function circleFrom2Points(
  p1: Location,
  p2: Location
): { center: { lat: number; lng: number }; radius: number } {
  const center = {
    lat: (p1.lat + p2.lat) / 2,
    lng: (p1.lng + p2.lng) / 2,
  };
  const radius = haversineDistance(p1.lat, p1.lng, p2.lat, p2.lng) / 2;
  return { center, radius };
}

/**
 * Compute circle from 3 points using circumcircle
 */
function circleFrom3Points(
  p1: Location,
  p2: Location,
  p3: Location
): { center: { lat: number; lng: number }; radius: number } | null {
  // Use simple approximation for small areas
  // Convert to planar coordinates (works well for small regions)
  const ax = p1.lng;
  const ay = p1.lat;
  const bx = p2.lng;
  const by = p2.lat;
  const cx = p3.lng;
  const cy = p3.lat;

  const d = 2 * (ax * (by - cy) + bx * (cy - ay) + cx * (ay - by));

  if (Math.abs(d) < 1e-10) {
    // Collinear points, fall back to 2-point circle
    return circleFrom2Points(p1, p2);
  }

  const ux =
    ((ax * ax + ay * ay) * (by - cy) +
      (bx * bx + by * by) * (cy - ay) +
      (cx * cx + cy * cy) * (ay - by)) /
    d;
  const uy =
    ((ax * ax + ay * ay) * (cx - bx) +
      (bx * bx + by * by) * (ax - cx) +
      (cx * cx + cy * cy) * (bx - ax)) /
    d;

  const center = { lat: uy, lng: ux };
  const radius = haversineDistance(center.lat, center.lng, p1.lat, p1.lng);

  return { center, radius };
}

/**
 * Check if point is inside circle (with tolerance)
 */
function isInsideCircle(
  point: Location,
  circle: { center: { lat: number; lng: number }; radius: number },
  tolerance: number = 0.01
): boolean {
  const dist = haversineDistance(point.lat, point.lng, circle.center.lat, circle.center.lng);
  return dist <= circle.radius * (1 + tolerance);
}

/**
 * Welzl's algorithm for Minimum Enclosing Circle
 * Expected O(n) time complexity
 */
function welzlRecursive(
  points: Location[],
  boundary: Location[],
  n: number
): { center: { lat: number; lng: number }; radius: number } | null {
  // Base cases
  if (n === 0 || boundary.length === 3) {
    if (boundary.length === 0) return null;
    if (boundary.length === 1)
      return { center: { lat: boundary[0].lat, lng: boundary[0].lng }, radius: 0 };
    if (boundary.length === 2) return circleFrom2Points(boundary[0], boundary[1]);
    return circleFrom3Points(boundary[0], boundary[1], boundary[2]);
  }

  // Pick a random point
  const idx = Math.floor(Math.random() * n);
  const p = points[idx];

  // Swap with last point
  [points[idx], points[n - 1]] = [points[n - 1], points[idx]];

  // Recursively compute circle without p
  const circle = welzlRecursive(points, boundary, n - 1);

  // If p is inside circle, return circle
  if (circle && isInsideCircle(p, circle)) {
    return circle;
  }

  // Otherwise, p must be on the boundary
  return welzlRecursive(points, [...boundary, p], n - 1);
}

/**
 * Compute Minimum Enclosing Circle using Welzl's algorithm
 * Handles edge cases: duplicates, collinear sets, 2-point/3-point circles
 */
export function computeMinimumEnclosingCircle(locations: Location[]): Circle | null {
  if (locations.length === 0) return null;
  if (locations.length === 1) {
    return {
      center: { lat: locations[0].lat, lng: locations[0].lng },
      radius: 100, // Minimum radius of 100m for single point
    };
  }

  // Create a copy to avoid mutating original array
  const points = [...locations];
  const circle = welzlRecursive(points, [], points.length);

  if (!circle) return null;

  // Ensure minimum radius of 100m
  return {
    center: circle.center,
    radius: Math.max(circle.radius, 100),
  };
}

/**
 * Expand circle by epsilon factor (default 10% = 0.1)
 */
export function expandCircle(circle: Circle, epsilon: number = 0.1): Circle {
  return {
    center: circle.center,
    radius: circle.radius * (1 + epsilon),
  };
}

File: server/server/app/services/algorithms.py
"""Server-side algorithm implementations for geometric calculations."""

import math
from typing import List, Tuple, Optional


def compute_centroid(locations: List[Tuple[float, float]]) -> Optional[Tuple[float, float]]:
    """
    Compute the spherical centroid of multiple lat/lng coordinates.

    Args:
        locations: List of (lat, lng) tuples

    Returns:
        (lat, lng) tuple of centroid, or None if empty
    """
    if not locations:
        return None

    if len(locations) == 1:
        return locations[0]

    # Convert to unit vectors
    x_sum = 0.0
    y_sum = 0.0
    z_sum = 0.0

    for lat, lng in locations:
        lat_rad = math.radians(lat)
        lng_rad = math.radians(lng)

        x_sum += math.cos(lat_rad) * math.cos(lng_rad)
        y_sum += math.cos(lat_rad) * math.sin(lng_rad)
        z_sum += math.sin(lat_rad)

    # Average
    n = len(locations)
    x_avg = x_sum / n
    y_avg = y_sum / n
    z_avg = z_sum / n

    # Normalize and convert back
    lng_rad = math.atan2(y_avg, x_avg)
    hyp = math.sqrt(x_avg * x_avg + y_avg * y_avg)
    lat_rad = math.atan2(z_avg, hyp)

    return (math.degrees(lat_rad), math.degrees(lng_rad))


def haversine_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Calculate the great circle distance between two points on Earth.

    Args:
        lat1, lng1: First point coordinates
        lat2, lng2: Second point coordinates

    Returns:
        Distance in kilometers
    """
    R = 6371.0  # Earth radius in km

    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    dlat = math.radians(lat2 - lat1)
    dlng = math.radians(lng2 - lng1)

    a = (math.sin(dlat / 2) ** 2 +
         math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlng / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return R * c


def compute_mec(locations: List[Tuple[float, float]]) -> Optional[Tuple[float, float, float]]:
    """
    Compute the Minimum Enclosing Circle using Welzl's algorithm.

    Args:
        locations: List of (lat, lng) tuples

    Returns:
        (center_lat, center_lng, radius_km) tuple, or None if empty
    """
    if not locations:
        return None

    if len(locations) == 1:
        return (locations[0][0], locations[0][1], 1.0)  # 1km minimum for visibility and search purposes

    # Welzl's algorithm implementation
    def welzl_helper(points: List[Tuple[float, float]], boundary: List[Tuple[float, float]]) -> Tuple[float, float, float]:
        # Base cases
        if not points or len(boundary) == 3:
            return make_circle_from_boundary(boundary)

        # Pick a random point
        p = points[0]
        remaining = points[1:]

        # Recursively compute MEC without p
        circle = welzl_helper(remaining, boundary)

        # If p is inside the circle, return it
        if is_inside_circle(p, circle):
            return circle

        # Otherwise, p must be on the boundary
        return welzl_helper(remaining, boundary + [p])

    def make_circle_from_boundary(boundary: List[Tuple[float, float]]) -> Tuple[float, float, float]:
        """Create minimum circle from boundary points."""
        if len(boundary) == 0:
            return (0.0, 0.0, 0.0)
        elif len(boundary) == 1:
            return (boundary[0][0], boundary[0][1], 1.0)  # 1km minimum for visibility
        elif len(boundary) == 2:
            # Circle with two points as diameter
            lat1, lng1 = boundary[0]
            lat2, lng2 = boundary[1]
            center_lat = (lat1 + lat2) / 2
            center_lng = (lng1 + lng2) / 2
            radius = haversine_distance(lat1, lng1, lat2, lng2) / 2
            return (center_lat, center_lng, max(radius, 1.0))  # 1km minimum for visibility
        else:
            # Circle with three points
            return make_circle_from_three_points(boundary[0], boundary[1], boundary[2])

    def make_circle_from_three_points(
        p1: Tuple[float, float],
        p2: Tuple[float, float],
        p3: Tuple[float, float]
    ) -> Tuple[float, float, float]:
        """Create circle from three points using circumcircle."""
        # Simplified: use centroid and max distance
        center_lat = (p1[0] + p2[0] + p3[0]) / 3
        center_lng = (p1[1] + p2[1] + p3[1]) / 3

        r1 = haversine_distance(center_lat, center_lng, p1[0], p1[1])
        r2 = haversine_distance(center_lat, center_lng, p2[0], p2[1])
        r3 = haversine_distance(center_lat, center_lng, p3[0], p3[1])

        radius = max(r1, r2, r3)
        return (center_lat, center_lng, max(radius, 1.0))  # 1km minimum for visibility

    def is_inside_circle(point: Tuple[float, float], circle: Tuple[float, float, float]) -> bool:
        """Check if point is inside circle."""
        center_lat, center_lng, radius = circle
        distance = haversine_distance(point[0], point[1], center_lat, center_lng)
        return distance <= radius * 1.001  # Small tolerance for floating point

    # Run Welzl's algorithm
    import random
    shuffled = locations.copy()
    random.shuffle(shuffled)
    return welzl_helper(shuffled, [])


def apply_fuzzing(lat: float, lng: float, radius_km: float = 0.5) -> Tuple[float, float]:
    """
    Apply random offset to coordinates for privacy (blur mode).

    Args:
        lat: Original latitude
        lng: Original longitude
        radius_km: Fuzzing radius in kilometers

    Returns:
        (fuzzy_lat, fuzzy_lng) tuple
    """
    import random

    # Convert km to degrees (approximate)
    lat_offset = (random.random() - 0.5) * 2 * (radius_km / 111.0)
    lng_offset = (random.random() - 0.5) * 2 * (radius_km / (111.0 * math.cos(math.radians(lat))))

    return (lat + lat_offset, lng + lng_offset)
